src/v2/providers/storage.ts

```diff
@@ -0,0 +1,389 @@
+  opts: StorageOptions,
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData>;
+
+export function onObjectFinalized(
+  buketOrOptsOrHandler:
+    | string
+    | StorageOptions
+    | ((event: CloudEvent<StorageObjectData>) => any | Promise<any>),
+  handler?: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData> {
+  return _onOperation(finalizedEvent, buketOrOptsOrHandler, handler);
+}
+
+/** Handle a storage object deleted */
+export function onObjectDeleted(
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData>;
+
+/** Handle a storage object deleted */
+export function onObjectDeleted(
+  bucket: string,
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData>;
+
+/** Handle a storage object deleted */
+export function onObjectDeleted(
+  opts: StorageOptions,
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData>;
+
+export function onObjectDeleted(
+  buketOrOptsOrHandler:
+    | string
+    | StorageOptions
+    | ((event: CloudEvent<StorageObjectData>) => any | Promise<any>),
+  handler?: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData> {
+  return _onOperation(deletedEvent, buketOrOptsOrHandler, handler);
+}
+
+/** Handle a storage object metadata updated */
+export function onObjectMetadataUpdated(
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData>;
+
+/** Handle a storage object metadata updated */
+export function onObjectMetadataUpdated(
+  bucket: string,
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData>;
+
+/** Handle a storage object metadata updated */
+export function onObjectMetadataUpdated(
+  opts: StorageOptions,
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData>;
+
+export function onObjectMetadataUpdated(
+  buketOrOptsOrHandler:
+    | string
+    | StorageOptions
+    | ((event: CloudEvent<StorageObjectData>) => any | Promise<any>),
+  handler?: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData> {
+  return _onOperation(metadataUpdatedEvent, buketOrOptsOrHandler, handler);
+}
+
+/** @hidden */
+export function _onOperation(
+  // handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>,
+  // eventType: string,
+  // bucketOrOpts?: string | StorageOptions
+  eventType: string,
+  bucketOrOptsOrHandler:
+    | string
+    | StorageOptions
+    | ((event: CloudEvent<StorageObjectData>) => any | Promise<any>),
+  handler?: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData> {
+  if (arguments.length === 2) {
```

Comment:
Isn't there a subtle bug here since `_onOperation` is always called with 3 arguments now?

e.g.
```js
> function foo(a, b) { return arguments.length }
undefined
> foo(1)
1
> foo(1, 2)
2
> foo(1, undefined)
2
```

So I'd expect `arguments.length` to always be `3`

Reply:
Maybe this will work:

```
if (typeof bucketOrOptsOrHandler === "function") {
  ...
}
```

Reply:
Oh dang, yeah that's a bug alright. For some reason in my head passing an undefined handler was equivalent to leaving off the last parameter. Thanks for catching this!